flink-cdc-composer/src/main/java/com/ververica/cdc/composer/utils/FactoryDiscoveryUtils.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2023 Ververica Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.ververica.cdc.composer.utils;
+
+import com.ververica.cdc.common.factories.Factory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ServiceLoader;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+/** Discovery utilities for {@link Factory}. */
+public class FactoryDiscoveryUtils {
+
+    private static final Logger LOG = LoggerFactory.getLogger(FactoryDiscoveryUtils.class);
+
+    private FactoryDiscoveryUtils() {}
+
+    /** Returns the {@link Factory} for the given identifier. */
+    @SuppressWarnings("unchecked")
+    public static <T extends Factory> T getFactoryByIdentifier(
+            String identifier, Class<T> implementClass) {
+
+        final ServiceLoader<Factory> loader = ServiceLoader.load(Factory.class);
+
+        final List<Factory> factoryList = new ArrayList<>();
+        final Iterator<Factory> factories = loader.iterator();
+        while (factories.hasNext()) {
+            try {
+                final Factory factory = factories.next();
+                if (factory != null
+                        && factory.identifier().equals(identifier)
+                        && implementClass.isAssignableFrom(factory.getClass())) {
+                    factoryList.add(factory);
+                }
+            } catch (Throwable e) {
+                if (e.getCause() instanceof NoClassDefFoundError) {
```

Comment:
I don't quite get the logic of handling `NoClassDefFoundError` here. The same logic in Flink is for handling a special case where `flink-connector-files` is missing. I think we don't have that issue in our project, so maybe just throw the exception directly?